internal/characters/xiao/xiao.go

```diff
@@ -0,0 +1,128 @@
+package xiao
+
+import (
+	"github.com/genshinsim/gsim/pkg/core"
+	"github.com/genshinsim/gsim/pkg/character"
+)
+
+func init() {
+	core.RegisterCharFunc("xiao", NewChar)
+}
+
+// Xiao specific character implementation
+type char struct {
+	*character.Tmpl
+	eCharge      int
+	eChargeMax   int
+	eNextRecover int
+	eTickSrc     int
+	qStarted     int
+	a4Expiry     int
+	c6Check      []interface{} // int (source frame), string (snapshot ability name)
+	c6Activated  int
+}
+
+// Initializes character
+// TODO: C4 is not implemented - don't really care about def
+func NewChar(s *core.Core, p core.CharacterProfile) (core.Character, error) {
+	c := char{}
+	t, err := character.NewTemplateChar(s, p)
+	if err != nil {
+		return nil, err
+	}
+	c.Tmpl = t
+	c.Energy = 70
+	c.EnergyMax = 70
+	c.Weapon.Class = core.WeaponClassSpear
+	c.BurstCon = 5
+	c.SkillCon = 3
+	c.NormalHitNum = 5
+
+	c.eChargeMax = 2
+	if c.Base.Cons >= 1 {
+		c.eChargeMax = 3
+	}
+
+	if c.Base.Cons >= 2 {
+		c.c2()
+	}
+	if c.Base.Cons >= 6 {
+		c.c6Check = append(c.c6Check, 0, "")
+		c.c6()
+	}
+
+	c.eCharge = c.eChargeMax
+	c.onExitField()
+
+	return &c, nil
+}
+
+// Implements Xiao C2:
+// When in the party and not on the field, Xiao's Energy Recharge is increased by 25%
+func (c *char) c2() {
+	stat_mod := make([]float64, core.EndStatType)
+	stat_mod[core.ER] = 0.25
+	c.AddMod(core.CharStatMod{
+		Key: "xiao-c2",
+		Expiry: -1,
+		Amount: func(a core.AttackTag) ([]float64, bool) {
+			if c.Core.ActiveChar != c.Index {
+				return stat_mod, true
+			}
+			return nil, false
+		},
+	})
+}
+
+// Implements Xiao C6:
+// While under the effect of Bane of All Evil, hitting at least 2 opponents with Xiao's Plunge Attack will immediately grant him 1 charge of Lemniscatic Wind Cycling, and for the next 1s, he may use Lemniscatic Wind Cycling while ignoring its CD.
+// Adds an OnDamage event checker - if we record two or more instances of plunge damage, then activate C6
+func (c *char) c6() {
+	c.Core.Events.Subscribe(core.OnDamage, func(args ...interface{}) bool {
+		ds := args[1].(*core.Snapshot)
+		if ds.ActorIndex != c.Index {
+			return false
+		}
+		if !((ds.Abil == "High Plunge") || (ds.Abil == "Low Plunge")) {
```

Comment:
Here can check for tag instead

```go
if ds.AttackTag != core.AttackTagPlunge {
   return false
}
if c.Core.Status.Duration("xiaoburst") == 0 {
    return false
}
```

since actor is already xiao and this is plunge then it should be fine. should also check if his burst is activated on the next line though?

Reply:
Can't use the plunge tag since you don't want the normal plunging hits to be counted towards the proc, as otherwise you would be able to proc C6 Xiao on single entities which we know isn't the case.